File: backend/app/games/baseball.py
"""Baseball: the House pitches on the server timeline; a swing's t_server is compared with the pitch's arrival.
A turn is one human at-bat (several pitches in one input window); after the third out the House bats its half."""
from __future__ import annotations

from typing import Any

from ..agents.pitcher import PitcherPolicy
from ..agents.policy import DecisionRequest, Option
from .base import Match, MarketSpec, TurnPlan, TurnResult

OUTCOME_ORDER = ["strikeout", "out", "foul", "single", "double", "home_run"]


def contact(dt_ms: float, W: float, swing_angle: float, pitch_height: float, power: float, rot_sign: int, rng) -> dict[str, Any]:
    """Spec §7.3: hit if |Δt| ≤ W; quality from timing and height; direction from timing; distance from power and quality."""
    if abs(dt_ms) > W:
        return {"result": "miss", "quality": 0.0, "dir": 0.0, "dist": 0.0, "dt_ms": round(dt_ms)}
    quality = (1 - abs(dt_ms) / W) * (1 - 0.6 * abs(swing_angle - pitch_height))
    direction = max(-1.0, min(1.0, -dt_ms / W + 0.3 * rot_sign))
    dist = max(0.0, min(1.0, power)) * (0.4 + 0.6 * quality)
    r = rng.random()
    if quality < 0.25:
        res = "foul" if r < 0.6 else "out"
    elif quality < 0.5:
        res = "out" if r < 0.55 else "single"
    elif quality < 0.75:
        res = "single" if r < 0.5 else ("double" if r < 0.8 else "out")
    else:
        res = "home_run" if dist > 0.8 else "double"
    return {"result": res, "quality": round(quality, 3), "dir": round(direction, 3), "dist": round(dist, 3), "dt_ms": round(dt_ms)}


class Side:
    def __init__(self, name: str):
        self.name = name
        self.runs = 0
        self.hits = 0
        self.outs = 0
        self.bases = [False, False, False]
        self.log: list[str] = []

    def advance(self, n: int) -> int:
        """Advance runners by n bases (batter included). Returns runs scored."""
        runs = 0
        runners = [True] + self.bases            # batter at "base 0"
        for i in range(len(runners) - 1, -1, -1):
            if runners[i]:
                dest = i + n
                if dest >= 4:
                    runs += 1
                else:
                    runners[dest] = True
                runners[i] = False if dest != i else runners[i]
        self.bases = runners[1:4]
        self.runs += runs
        return runs

    def new_inning(self) -> None:
        self.outs = 0
        self.bases = [False, False, False]

    def public(self) -> dict[str, Any]:
        return {"name": self.name, "runs": self.runs, "hits": self.hits, "outs": self.outs, "bases": list(self.bases)}


class BaseballMatch(Match):
    sport = "baseball"

    def __init__(self, *a, **kw):
        super().__init__(*a, **kw)
        g = self.config.get("game.baseball")
        self.innings, self.outs_per_half = int(g["innings"]), int(g["outs_per_half"])
        self.pitch_defs: dict[str, dict[str, float]] = dict(g["pitches"])
        self.policy = PitcherPolicy(self.tier.params)
        self.human = Side(", ".join(self.players.values()) or "Player")
        self.house = Side(self.tier.name)
        self.inning = 1
        self.sudden_death = bool(self.tier.params.get("sudden_death", False))
        self.scenario_pitches = list(self.scenario.get("pitches", []))
        self.pitch_no = 0
        self.pitch: dict[str, Any] | None = None      # the pitch in flight (public fields)
        self.arrival_ts: float | None = None            # nominal arrival, ms server time
        self.actual_arrival: float | None = None        # with late break
        self.next_release_at: float | None = None
        self.strikes = 0
        self.at_bat_over = False
        self.at_bat_result: dict[str, Any] | None = None
        self.swung = False
        self.events: list[dict[str, Any]] = []
        self.house_ticks: list[dict[str, Any]] = []
        self.pitch_gap_s = float(g.get("pitch_gap_s", 2.5))
        self.last_swing_dt: float | None = None
        self.dts: list[float] = []

    # ---- turn = one human at-bat ------------------------------------------------------------------
    def plan_turn(self) -> TurnPlan | None:
        if self.inning > self.innings or (self.sudden_death and self.at_bat_result and self.at_bat_result.get("sudden_death_over")):
            return None
        self.strikes, self.at_bat_over, self.at_bat_result, self.pitch, self.arrival_ts, self.swung = 0, False, None, None, None, False
        self.next_release_at = None
        self.events, self.house_ticks = [], []
        label = f"Inning {self.inning} · {self.human.name} at bat"
        plan = TurnPlan(label=label, input_seats=list(self.seats), input_window_s=float(self.config.get("game.baseball.at_bat_window_s", 45)),
                        markets=[MarketSpec("turn_outcome", f"{label}", [("hit", "Hit"), ("out", "Out")])],
                        prompt={"inning": self.inning, "outs": self.human.outs, "bases": list(self.human.bases), "sudden_death": self.sudden_death})
        plan.tick_ms = 100.0  # type: ignore[attr-defined]
        return plan

    def decisions_before(self, phase: str) -> list[DecisionRequest]:
        if phase != "input":
            return []
        state = {"inning": self.inning, "human_runs": self.human.runs, "house_runs": self.house.runs, "outs": self.human.outs,
                 "mean_dt_ms": self.policy.mean_dt, "swing_rate": round(self.policy.swing_rate, 2)}
        opts = self.policy.options(state, self.rng)
        return [DecisionRequest(agent_id=f"house:baseball:{self.tier.id}", options=opts, default=self.policy.default(opts, self.rng), context=state, turn_no=self.turn_no, match_id=self.match_id)]

    def apply_decision(self, req: DecisionRequest, option: Option) -> None:
        self.policy.mode = str(option.params.get("mode", "mix"))

    def on_gesture(self, seat_id: str, gesture) -> bool:
        if gesture.kind != "swing" or self.at_bat_over:
            return False
        if self.pitch is None or self.actual_arrival is None or self.swung:
            self.events.append({"kind": "swing", "result": "early", "no_pitch": True})
            return True
        self.swung = True
        dt = float(gesture.t_server) - float(self.actual_arrival)
        W = self.param("W_ms", 90)
        angle = float(gesture.extra.get("pitch_angle", 0.5))
        rot = int(gesture.sign or 1)
        c = contact(dt, W, angle, float(self.pitch["height"]), float(gesture.power), rot, self.rng)
        self.policy.observe(dt)
        self.last_swing_dt = dt
        self.dts.append(dt)
        self._resolve_pitch(c, swung=True)
        return True

    def input_done(self) -> bool:
        return self.at_bat_over

    def no_input(self) -> None:
        if not self.at_bat_over:
            self.at_bat_result = {"result": "strikeout", "reason": "no_swing"}
            self.human.outs += 1
            self.at_bat_over = True

    def step(self, dt_s: float | None = None, now: float | None = None) -> dict[str, Any]:
        """Called every 100 ms during the at-bat: releases pitches and calls missed ones."""
        self.events = []
        if now is None or self.at_bat_over:
            return self.tick_summary(now)
        now_ms = now * 1000
        if self.pitch is None:
            if self.next_release_at is None:
                self.next_release_at = now_ms + 1500
            if now_ms >= self.next_release_at:
                self._release(now_ms)
        else:
            W = self.param("W_ms", 90)
            if now_ms > float(self.actual_arrival) + W + 60 and not self.swung:
                self.policy.observe(None)
                self._resolve_pitch({"result": "called", "quality": 0, "dir": 0, "dist": 0, "dt_ms": None}, swung=False)
        return self.tick_summary(now)

    def _release(self, now_ms: float) -> None:
        self.pitch_no += 1
        if self.scenario_pitches:
            kind = self.scenario_pitches[(self.pitch_no - 1) % len(self.scenario_pitches)]
        else:
            kind = self.policy.choose_pitch(self.rng)
        pd = self.pitch_defs.get(kind, self.pitch_defs["fastball"])
        speed_mult = self.param("speed_mult", 1.0)
        travel = float(pd["travel_ms"]) / max(0.5, speed_mult)
        late = self.rng.uniform(-90, 90) if pd.get("late_break") else 0.0
        self.arrival_ts = now_ms + travel
        self.actual_arrival = self.arrival_ts + late
        self.pitch = {"kind": kind, "release_ts": int(now_ms), "arrival_ts": int(self.arrival_ts), "travel_ms": int(travel), "height": float(pd["height"]),
                      "brk": float(pd["brk"]), "late_break_ms": int(late), "no": self.pitch_no, "strikes": self.strikes}
        self.swung = False
        self.events.append({"kind": "pitch", **self.pitch})

    def _resolve_pitch(self, c: dict[str, Any], swung: bool) -> None:
        res = c["result"]
        ev = {"kind": "result", "pitch_no": self.pitch_no, "swung": swung, **c}
        if res in ("miss", "called"):
            self.strikes += 1
            ev["strikes"] = self.strikes
            if self.strikes >= 3:
                self.at_bat_result = {"result": "strikeout", "reason": "swinging" if swung else "looking"}
                self.human.outs += 1
                self.at_bat_over = True
        elif res == "foul":
            if self.strikes < 2:
                self.strikes += 1
            ev["strikes"] = self.strikes
        elif res == "out":
            self.at_bat_result = {"result": "out", **c}
            self.human.outs += 1
            self.at_bat_over = True
        else:
            bases = {"single": 1, "double": 2, "home_run": 4}[res]
            runs = self.human.advance(bases)
            self.human.hits += 1
            self.at_bat_result = {"result": res, "runs": runs, **c}
            self.at_bat_over = True
        self.events.append(ev)
        if not self.at_bat_over:
            gap_from = float(self.actual_arrival or 0)
            self.pitch, self.arrival_ts, self.actual_arrival = None, None, None
            self.next_release_at = gap_from + self.pitch_gap_s * 1000
        else:
            self.pitch = None

    def tick_summary(self, now: float | None = None) -> dict[str, Any]:
        d = {"inning": self.inning, "batter": self.seats[0] if self.seats else None, "pitch": self.pitch, "pitch_no": self.pitch_no, "arrival_ts": self.arrival_ts,
             "strikes": self.strikes, "human": self.human.public(), "house": self.house.public(), "events": list(self.events), "at_bat_over": self.at_bat_over,
             "sudden_death": self.sudden_death, "flags": {}}
        for e in self.events:
            if e.get("kind") == "result" and e.get("result") == "home_run":
                d["flags"] = {"shake": True, "hitstop": True}
            elif e.get("kind") == "result" and e.get("result") in ("single", "double"):
                d["flags"] = {"shake": True}
        return d

    # ---- resolution: settle the at-bat; after 3 outs the House bats ---------------------------------------------
    def resolve(self) -> TurnResult:
        r = self.at_bat_result or {"result": "strikeout"}
        outcome = r["result"]
        hit = outcome in ("single", "double", "home_run")
        ticks: list[dict[str, Any]] = []
        triggers: list[tuple[str, dict[str, Any]]] = []
        player = self.human.name
        if outcome == "home_run":
            triggers.append(("home_run", {"player": player}))
        elif hit:
            triggers.append(("turn_result", {"outcome": outcome.replace("_", " ").title(), "player": player}))
        else:
            triggers.append(("taunt", {"outcome": outcome}))
        sudden_over = False
        if self.sudden_death:                     # The Closer: any out ends it for the House; a run wins it for the human
            if not hit:
                sudden_over = True
                self.house.runs = max(self.house.runs, self.human.runs + 1)
            elif r.get("runs", 0) > 0:
                sudden_over = True
        if self.human.outs >= self.outs_per_half and not sudden_over:
            ticks.extend(self._house_half())
            self.human.new_inning()
            self.inning += 1
        anim = 2.2 + sum(t.get("anim_s", 1.5) for t in ticks)
        detail = {"at_bat": r, "inning": self.inning, "sudden_death_over": sudden_over, "house_half": [t.get("result") for t in ticks]}
        if sudden_over:
            self.at_bat_result = dict(r, sudden_death_over=True)
            self.inning = self.innings + 1
        return TurnResult(outcome="hit" if hit else "out", detail=detail, score=self.summary(), market_winners={"turn_outcome": ["hit" if hit else "out"]},
                          triggers=triggers, animation_s=anim, ticks=ticks)

    def _house_half(self) -> list[dict[str, Any]]:
        """The House bats: each at-bat resolves from the tier's timing distribution."""
        ticks = []
        self.house.new_inning()
        sigma = float(self.tier.params.get("house_dt_sigma_ms", 60))
        W = 90.0
        n = 0
        while self.house.outs < self.outs_per_half and n < 12:
            n += 1
            dt = self.rng.gauss(0, sigma)
            c = contact(dt, W, 0.5, self.rng.uniform(0.3, 0.7), self.rng.uniform(0.5, 1.0), 1, self.rng)
            res = c["result"]
            runs = 0
            if res in ("miss", "foul", "out"):
                self.house.outs += 1
                res = "out"
            else:
                runs = self.house.advance({"single": 1, "double": 2, "home_run": 4}[res])
                self.house.hits += 1
            ticks.append({"who": "house", "result": res, "runs": runs, "dir": c["dir"], "dist": c["dist"], "house": self.house.public(), "anim_s": 1.6 if res != "home_run" else 2.4,
                          "flags": {"shake": res == "home_run"}})
        return ticks

    def summary(self) -> dict[str, Any]:
        return {"inning": min(self.inning, self.innings), "innings": self.innings, "human": self.human.public(), "house": self.house.public(),
                "sudden_death": self.sudden_death, "mean_dt_ms": round(self.policy.mean_dt, 1) if self.policy.mean_dt is not None else None}

    def winner(self) -> str:
        if self.human.runs != self.house.runs:
            return "human" if self.human.runs > self.house.runs else "house"
        return "tie" if not self.sudden_death else "house"

File: web/src/game/sims/baseball.ts
// Baseball: the House pitches on the client clock; a swing's time is compared to the pitch's arrival (backend/app/games/baseball.py port).
import { Rng } from './rng'
import type { DecisionReq, Gesture, Sim, StartPayload, TurnPlan, TurnResult } from './types'

export interface Side { runs: number; hits: number; outs: number; bases: boolean[] }
export interface Pitch { kind: string; release_ts: number; arrival_ts: number; travel_ms: number; height: number; brk: number; late_break_ms: number; no: number; strikes: number }
export interface BbEvent { kind: string; result?: string; strikes?: number; dt_ms?: number | null; dir?: number; dist?: number; quality?: number; swung?: boolean; pitch_no?: number; no_pitch?: boolean; pitch?: Pitch }
const OFFSPEED = ['changeup', 'curve', 'knuckle'], LOCATION = ['high', 'low', 'inside', 'outside']

export class BaseballSim implements Sim {
  readonly sport = 'baseball'
  rng = new Rng(1)
  start!: StartPayload
  params: Record<string, unknown> = {}
  adjustments: Record<string, number> = {}
  innings = 3; outsPerHalf = 3; pitchGapS = 2.5; atBatWindowS = 45
  pitchDefs: Record<string, { travel_ms: number; brk: number; height: number; late_break?: number }> = {}
  human: Side = { runs: 0, hits: 0, outs: 0, bases: [false, false, false] }
  house: Side = { runs: 0, hits: 0, outs: 0, bases: [false, false, false] }
  inning = 1; suddenDeath = false
  pitch: Pitch | null = null; pitchNo = 0; arrivalMs = -1; actualArrival = -1; nextReleaseMs = -1
  strikes = 0; atBatOver = false; atBatResult: Record<string, unknown> = {}; swung = false
  events: BbEvent[] = []
  dts: number[] = []; swingN = 0; swingS = 0; pmode = 'mix'
  seats: string[] = []
  private plan: TurnPlan | null = null
  turnNo = 0

  setup(start: StartPayload): void {
    this.start = start; this.rng = new Rng(start.seed); this.params = { ...start.tier.params }
    const g = (start.game.baseball ?? {}) as Record<string, unknown>
    this.innings = Number(g.innings ?? 3); this.outsPerHalf = Number(g.outs_per_half ?? 3); this.pitchGapS = Number(g.pitch_gap_s ?? 2.5); this.atBatWindowS = Number(g.at_bat_window_s ?? 45)
    this.pitchDefs = (g.pitches ?? {}) as typeof this.pitchDefs
    this.suddenDeath = !!this.params.sudden_death
    this.seats = start.seats.map((s) => s.seat_id)
    for (const [k, v] of Object.entries(start.adjustments ?? {})) this.adjust(k, v, 'adaptive')
  }
  param(n: string, d: number): number { return Number(this.params[n] ?? d) + (this.adjustments[n] ?? 0) }
  adjust(name: string, delta: number, reason: string): void { this.adjustments[name] = reason === 'adaptive' ? delta : (this.adjustments[name] ?? 0) + delta }
  grant(): void {}
  humanLabel(): string { return Object.values(this.start.players).join(', ') || 'Player' }

  planTurn(): TurnPlan | null {
    if (this.inning > this.innings || this.atBatResult.sudden_death_over) return null
    this.strikes = 0; this.atBatOver = false; this.atBatResult = {}; this.pitch = null; this.arrivalMs = -1; this.swung = false; this.nextReleaseMs = -1; this.events = []
    const label = `Inning ${this.inning} · ${this.humanLabel()} at bat`
    this.plan = { label, input_seats: [this.seats[0]], input_window_s: this.atBatWindowS, betting: true, live: true,
      markets: [{ kind: 'turn_outcome', label, outcomes: [['hit', 'Hit'], ['out', 'Out']] }],
      prompt: { inning: this.inning, outs: this.human.outs, bases: this.human.bases, sudden_death: this.suddenDeath } }
    return this.plan
  }
  currentPlan(): TurnPlan | null { return this.plan }
  meanDt(): number { return this.dts.length ? this.dts.reduce((a, b) => a + b, 0) / this.dts.length : 0 }
  swingRate(): number { return this.swingN ? this.swingS / this.swingN : 1 }
  decisionsBefore(phase: 'input' | 'resolving'): DecisionReq[] {
    if (phase !== 'input') return []
    const md = this.meanDt()
    return [{ agent_id: `house:baseball:${this.start.tier.id}`, default: 'attack', context: { inning: this.inning, human_runs: this.human.runs, house_runs: this.house.runs, mean_dt_ms: md },
      options: [{ id: 'attack', label: 'Attack with fastballs', ev: 0.45 }, { id: 'paint', label: 'Paint the corners', ev: this.swingRate() < 0.6 ? 0.5 : 0.4 }, { id: 'deceive', label: 'Change speeds', ev: this.dts.length && Math.abs(md) > 25 ? 0.55 : 0.42 }] }]
  }
  applyDecision(_r: DecisionReq, id: string): void { this.pmode = id }
  choosePitch(): string {
    const allowed = (this.params.pitches as string[] | undefined) ?? ['fastball', 'changeup', 'curve'], read = Number(this.params.read_strength ?? 0.5), md = this.meanDt()
    if (this.rng.next() < read && this.dts.length >= 2) {
      if (md < -30) { const c = allowed.filter((p) => OFFSPEED.includes(p)); if (c.length) return this.rng.choice(c) }
      else if (md > 30 && allowed.includes('fastball')) return 'fastball'
    }
    if (this.rng.next() < read && this.swingN >= 3 && this.swingRate() < 0.5) { const c = allowed.filter((p) => LOCATION.includes(p)); if (c.length) return this.rng.choice(c) }
    const w = allowed.map((p) => this.pmode === 'attack' ? (p === 'fastball' ? 3 : 0.6) : this.pmode === 'paint' ? (LOCATION.includes(p) ? 2.5 : 0.8) : this.pmode === 'deceive' ? (OFFSPEED.includes(p) ? 2.5 : 0.8) : 1)
    let r = this.rng.next() * w.reduce((a, b) => a + b, 0)
    for (let i = 0; i < allowed.length; i++) { r -= w[i]; if (r <= 0) return allowed[i] }
    return allowed[allowed.length - 1]
  }
  step(_dt: number, nowMs: number): void {
    this.events = []
    if (this.atBatOver) return
    if (!this.pitch) { if (this.nextReleaseMs < 0) this.nextReleaseMs = nowMs + 1500; if (nowMs >= this.nextReleaseMs) this.release(nowMs) }
    else if (nowMs > this.actualArrival + this.param('W_ms', 90) + 60 && !this.swung) { this.swingN++; this.resolvePitch({ result: 'called', quality: 0, dir: 0, dist: 0, dt_ms: null }, false) }
  }
  private release(nowMs: number): void {
    this.pitchNo++
    const sp = (this.start.scenario?.pitches as string[] | undefined) ?? []
    const kind = sp.length ? sp[(this.pitchNo - 1) % sp.length] : this.choosePitch()
    const pd = this.pitchDefs[kind] ?? this.pitchDefs.fastball ?? { travel_ms: 900, brk: 0, height: 0.5 }
    const travel = pd.travel_ms / Math.max(0.5, this.param('speed_mult', 1)), late = pd.late_break ? this.rng.range(-90, 90) : 0
    this.arrivalMs = nowMs + travel; this.actualArrival = this.arrivalMs + late
    this.pitch = { kind, release_ts: Math.round(nowMs), arrival_ts: Math.round(this.arrivalMs), travel_ms: Math.round(travel), height: pd.height, brk: pd.brk, late_break_ms: Math.round(late), no: this.pitchNo, strikes: this.strikes }
    this.swung = false
    this.events.push({ kind: 'pitch', pitch: this.pitch })
  }
  contact(dt: number, W: number, angle: number, height: number, power: number, rot: number): { result: string; quality: number; dir: number; dist: number; dt_ms: number } {
    if (Math.abs(dt) > W) return { result: 'miss', quality: 0, dir: 0, dist: 0, dt_ms: Math.round(dt) }
    const quality = (1 - Math.abs(dt) / W) * (1 - 0.6 * Math.abs(angle - height)), dir = Math.max(-1, Math.min(1, -dt / W + 0.3 * rot)), dist = Math.max(0, Math.min(1, power)) * (0.4 + 0.6 * quality), r = this.rng.next()
    const result = quality < 0.25 ? (r < 0.6 ? 'foul' : 'out') : quality < 0.5 ? (r < 0.55 ? 'out' : 'single') : quality < 0.75 ? (r < 0.5 ? 'single' : r < 0.8 ? 'double' : 'out') : dist > 0.8 ? 'home_run' : 'double'
    return { result, quality: +quality.toFixed(3), dir: +dir.toFixed(3), dist: +dist.toFixed(3), dt_ms: Math.round(dt) }
  }
  onGesture(seat: string, g: Gesture): boolean {
    if (g.kind !== 'swing' || this.atBatOver || seat !== this.seats[0]) return false
    if (!this.pitch || this.swung) { this.events.push({ kind: 'swing', result: 'early', no_pitch: true }); return true }
    this.swung = true
    const dt = g.t_server - this.actualArrival
    const c = this.contact(dt, this.param('W_ms', 90), Number(g.extra?.pitch_angle ?? 0.5), this.pitch.height, g.power, 1)
    this.swingN++; this.swingS++; this.dts.push(dt); if (this.dts.length > 6) this.dts.shift()
    this.resolvePitch(c, true); return true
  }
  private advance(side: Side, n: number): number {
    let runs = 0; const runners = [true, ...side.bases]
    for (let i = runners.length - 1; i >= 0; i--) if (runners[i]) { const dest = i + n; if (dest >= 4) runs++; else runners[dest] = true; if (dest !== i) runners[i] = false }
    side.bases = runners.slice(1, 4); side.runs += runs; return runs
  }
  private resolvePitch(c: { result: string; quality: number; dir: number; dist: number; dt_ms: number | null }, swung: boolean): void {
    const res = c.result, ev: BbEvent = { kind: 'result', pitch_no: this.pitchNo, swung, ...c }
    if (res === 'miss' || res === 'called') { this.strikes++; ev.strikes = this.strikes; if (this.strikes >= 3) { this.atBatResult = { result: 'strikeout', reason: swung ? 'swinging' : 'looking' }; this.human.outs++; this.atBatOver = true } }
    else if (res === 'foul') { if (this.strikes < 2) this.strikes++; ev.strikes = this.strikes }
    else if (res === 'out') { this.atBatResult = { ...c, result: 'out' }; this.human.outs++; this.atBatOver = true }
    else { const runs = this.advance(this.human, { single: 1, double: 2, home_run: 4 }[res] ?? 1); this.human.hits++; this.atBatResult = { ...c, runs }; this.atBatOver = true }
    this.events.push(ev)
    if (!this.atBatOver) this.nextReleaseMs = this.actualArrival + this.pitchGapS * 1000
    this.pitch = null; this.arrivalMs = -1
  }
  inputDone(): boolean { return this.atBatOver }
  noInput(): void { if (!this.atBatOver) { this.atBatResult = { result: 'strikeout', reason: 'no_swing' }; this.human.outs++; this.atBatOver = true } }
  resolve(): TurnResult {
    const r = Object.keys(this.atBatResult).length ? this.atBatResult : { result: 'strikeout' }
    const outcome = String(r.result), hit = ['single', 'double', 'home_run'].includes(outcome)
    const triggers: [string, Record<string, unknown>][] = outcome === 'home_run' ? [['home_run', { player: this.humanLabel() }]] : hit ? [['turn_result', { outcome: outcome.replace('_', ' '), player: this.humanLabel() }]] : [['taunt', { outcome }]]
    let suddenOver = false
    if (this.suddenDeath) { if (!hit) { suddenOver = true; this.house.runs = Math.max(this.house.runs, this.human.runs + 1) } else if (Number(r.runs ?? 0) > 0) suddenOver = true }
    let ticks: Record<string, unknown>[] = []
    if (this.human.outs >= this.outsPerHalf && !suddenOver) { ticks = this.houseHalf(); this.human.outs = 0; this.human.bases = [false, false, false]; this.inning++ }
    const anim = 2.2 + ticks.reduce((a, t) => a + Number(t.anim_s ?? 1.5), 0)
    if (suddenOver) { this.atBatResult.sudden_death_over = true; this.inning = this.innings + 1 }
    return { outcome: hit ? 'hit' : 'out', market_winners: { turn_outcome: [hit ? 'hit' : 'out'] }, detail: { at_bat: r, inning: this.inning, sudden_death_over: suddenOver }, triggers, animation_s: anim, ticks }
  }
  private houseHalf(): Record<string, unknown>[] {
    const ticks: Record<string, unknown>[] = []; this.house.outs = 0; this.house.bases = [false, false, false]
    const sigma = Number(this.params.house_dt_sigma_ms ?? 60); let n = 0
    while (this.house.outs < this.outsPerHalf && n < 12) {
      n++
      const c = this.contact(this.rng.gauss(0, sigma), 90, 0.5, this.rng.range(0.3, 0.7), this.rng.range(0.5, 1), 1)
      let res = c.result, runs = 0
      if (['miss', 'foul', 'out'].includes(res)) { this.house.outs++; res = 'out' } else { runs = this.advance(this.house, { single: 1, double: 2, home_run: 4 }[res] ?? 1); this.house.hits++ }
      ticks.push({ who: 'house', result: res, runs, dir: c.dir, dist: c.dist, house: { ...this.house, bases: [...this.house.bases] }, anim_s: res === 'home_run' ? 2.4 : 1.6 })
    }
    return ticks
  }
  summary(): Record<string, unknown> { return { inning: Math.min(this.inning, this.innings), innings: this.innings, human: { ...this.human, bases: [...this.human.bases] }, house: { ...this.house, bases: [...this.house.bases] }, sudden_death: this.suddenDeath, mean_dt_ms: +this.meanDt().toFixed(1) } }
  stateSummary(): Record<string, unknown> { return { score: this.summary(), pitch: this.pitch, pitch_no: this.pitchNo, arrival_ts: this.arrivalMs, strikes: this.strikes, batter: this.seats[0], at_bat_over: this.atBatOver } }
  winner(): string { if (this.human.runs !== this.house.runs) return this.human.runs > this.house.runs ? 'human' : 'house'; return this.suddenDeath ? 'house' : 'tie' }
}
